src/cryptography/hazmat/backends/openssl/backend.py

```diff
@@ -727,11 +727,17 @@ def create_x509_csr(self, builder, private_key, algorithm):
             isinstance(private_key, rsa.RSAPrivateKey)
         ):
             raise ValueError(
-                "MD5 is not a supported hash algorithm for EC/DSA CSRs"
+                "MD5 is only (reluctantly) supported for RSA CSRs"
             )
 
         # Resolve the signature algorithm.
-        evp_md = self._evp_md_non_null_from_algorithm(algorithm)
+        if isinstance(private_key, ed25519.Ed25519PrivateKey):
+            if not isinstance(algorithm, hashes.SHA512):
+                raise ValueError("ed25519 requires SHA512 hash algorithm")
+
+            evp_md = self._ffi.NULL
```

Comment:
This should have a comment explaining that openssl requires null for ed25519 at this time as it always does sha512